tss-esapi/src/structures/pcr/slot.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{tss2_esys::TPM2_PCR_SELECT_MAX, Error, Result, WrapperErrorKind};
+use enumflags2::{BitFlags, _internal::RawBitFlags};
+use log::error;
+use std::convert::TryFrom;
+
+/// Enum with the bit flag for each PCR slot.
+#[derive(BitFlags, Hash, Debug, Eq, PartialEq, Ord, PartialOrd, Clone, Copy)]
+#[repr(u32)]
+pub enum PcrSlot {
+    Slot0 = 0x0000_0001,
+    Slot1 = 0x0000_0002,
+    Slot2 = 0x0000_0004,
+    Slot3 = 0x0000_0008,
+    Slot4 = 0x0000_0010,
+    Slot5 = 0x0000_0020,
+    Slot6 = 0x0000_0040,
+    Slot7 = 0x0000_0080,
+    Slot8 = 0x0000_0100,
+    Slot9 = 0x0000_0200,
+    Slot10 = 0x0000_0400,
+    Slot11 = 0x0000_0800,
+    Slot12 = 0x0000_1000,
+    Slot13 = 0x0000_2000,
+    Slot14 = 0x0000_4000,
+    Slot15 = 0x0000_8000,
+    Slot16 = 0x0001_0000,
+    Slot17 = 0x0002_0000,
+    Slot18 = 0x0004_0000,
+    Slot19 = 0x0008_0000,
+    Slot20 = 0x0010_0000,
+    Slot21 = 0x0020_0000,
+    Slot22 = 0x0040_0000,
+    Slot23 = 0x0080_0000,
+}
+
+impl From<PcrSlot> for u32 {
+    fn from(pcr_slot: PcrSlot) -> u32 {
+        pcr_slot.bits()
+    }
+}
+
+impl TryFrom<u32> for PcrSlot {
+    type Error = Error;
+
+    fn try_from(pcr_slot: u32) -> Result<PcrSlot> {
+        BitFlags::<PcrSlot>::try_from(pcr_slot)
+            .map_err(|e| {
+                error!("Failed to convert tss pcr slot data to PcrSlot: {:?}", e);
+                Error::local_error(WrapperErrorKind::InvalidParam)
+            })
+            .and_then(|bit_flags| {
+                let mut pcr_slots_iter = bit_flags.iter();
+
+                pcr_slots_iter
+                    .next()
+                    .ok_or_else(|| {
+                        error!("No valid bit was set in the PcrSlot data");
+                        Error::local_error(WrapperErrorKind::InvalidParam)
+                    })
+                    .and_then(|pcr_slot| {
+                        if pcr_slots_iter.next().is_none() {
+                            Ok(pcr_slot)
+                        } else {
+                            error!("tss pcr slot data contained more then one value");
+                            Err(Error::local_error(WrapperErrorKind::InvalidParam))
+                        }
+                    })
```

Comment:
Oh this is old code that I just moved. Let me see if I can recall the reason behind my own madness. The easiest way would be to just do a count on the iterator!

It is not possible to do it the same way because this enum contains bitflags. So a single u32 value can contain several PcrSlots. I guess I could add FromPrimitive and ToPrimitive to this enum as well and then do the bitflag parsing else where.

Reply:
I changed this to give a more general error message.